bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataClientWrapper.java

```diff
@@ -46,39 +60,112 @@ public BigtableDataClientWrapper(BigtableDataClient bigtableDataClient,
+    ReadModifyWriteRowResponse response =
+        delegate.readModifyWriteRow(readModifyWriteRow.toProto(requestContext));
+    return convert(response.getRow());
   }
 
+  /** {@inheritDoc} */
   @Override
-  public ApiFuture<Row> readModifyWriteRowAsync(ReadModifyWriteRow readModifyWriteRow) {
-    throw new UnsupportedOperationException("Not implemented yet");
+  public ListenableFuture<Row> readModifyWriteRowAsync(ReadModifyWriteRow readModifyWriteRow) {
+    ListenableFuture<ReadModifyWriteRowResponse> response =
+        delegate.readModifyWriteRowAsync(readModifyWriteRow.toProto(requestContext));
+    return Futures.transform(response, new Function<ReadModifyWriteRowResponse, Row>() {
+      @NullableDecl
+      @Override
+      public Row apply(@NullableDecl ReadModifyWriteRowResponse input) {
+        return convert(input.getRow());
+      }
+    }, MoreExecutors.directExecutor());
   }
 
+  /** {@inheritDoc} */
   @Override
   public IBulkMutation createBulkMutationBatcher() {
     throw new UnsupportedOperationException("Not implemented yet");
   }
 
+  /** {@inheritDoc} */
   @Override
-  public ApiFuture<Boolean> checkAndMutateRowAsync(ConditionalRowMutation conditionalRowMutation) {
-    throw new UnsupportedOperationException("Not implemented yet");
+  public ListenableFuture<Boolean> checkAndMutateRowAsync(
+      ConditionalRowMutation conditionalRowMutation) {
+    final CheckAndMutateRowRequest request = conditionalRowMutation.toProto(requestContext);
+    final ListenableFuture<CheckAndMutateRowResponse> response =
+        delegate.checkAndMutateRowAsync(request);
+    return Futures.transform(response, new Function<CheckAndMutateRowResponse, Boolean>() {
+      @NullableDecl
+      @Override
+      public Boolean apply(@NullableDecl CheckAndMutateRowResponse input) {
+        return wasMutationApplied(request, input);
+      }
+    }, MoreExecutors.directExecutor());
   }
 
+  /** {@inheritDoc} */
   @Override
   public Boolean checkAndMutateRow(ConditionalRowMutation conditionalRowMutation) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    CheckAndMutateRowRequest request = conditionalRowMutation.toProto(requestContext);
+    CheckAndMutateRowResponse response =
+        delegate.checkAndMutateRow(conditionalRowMutation.toProto(requestContext));
+    return wasMutationApplied(request, response);
+  }
+
+  /**
+   * This method is referred from CheckAndMutateUtil#wasMutationApplied.
+   */
+  private static boolean wasMutationApplied(CheckAndMutateRowRequest request,
+      CheckAndMutateRowResponse response) {
+
+    // If we have true mods, we want the predicate to match.
+    // If we have false mods, we do not want the predicate to match.
+    return (request.getTrueMutationsCount() > 0
+        && response.getPredicateMatched()) || (
+        request.getFalseMutationsCount() > 0
+            && !response.getPredicateMatched());
```

Comment:
This looks odd. Why do we need to transform it? It becomes ambiguous when a conditional mutation has both true & false mutations

Reply:
If predicate matches for true mutations then we would not look for false mutation, as current implementation for [grpc.BigtableDataClient#checkAndMutateRow](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/1d1636f898cbde475c279d2673bc7efa4c503309/bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/BigtableDataClient.java#L88) is following this on [hbase1.x](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/36216cc9f2c5f9a71e361501c4d32b2c7f078c14/bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/AbstractBigtableTable.java#L459) & [hbase2.x](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/blob/82b9bb6db4e861161f164cd5e514ec2d8f9e96c8/bigtable-hbase-2.x-parent/bigtable-hbase-2.x/src/main/java/com/google/cloud/bigtable/hbase2_x/BigtableAsyncTable.java#L238).

I have just looked up GAX impl for the [checkAndMutate](https://github.com/googleapis/google-cloud-java/blob/bed08ffd27d1d232f229df20bc9f01da322131a8/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/CheckAndMutateRowCallable.java#L42) and now considering predicate match stautus.

Reply:
@igorbernstein2, All we care about is whether or not there was a mutation applied.  If there was a true mutation that was applied, then we're done; there's no need to check false mutations at that point.  Is there an objection to this?